### PR TITLE
Add a YAML MetaFactory

### LINKED ISSUE #80941: Add a YAML MetaFactory serialise multi-doc files
**What would you like to be added**:
Add an external MetaFactory that allows for serializing multiple YAML (or JSON) files in kubeadm. The logic of marshalling YAML should be factored out from `cmd/kubeadm/app/util/marshal` to `staging/src/k8s.io/apimachinery/pkg/runtime/serializer/`.

**Why is this needed**:
This will allow processing component configs containing multiple documents per file. Additionally, an out-of-tree library would allow sharing this functionality with other components.

### PR DESCRIPTION
**Pull request by obitech**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add an external MetaFactory that allows for serializing multiple YAML (or JSON) files in kubeadm. 

**Which issue(s) this PR fixes**:

Fixes #80941

**Special notes for your reviewer**:
I'm carrying over @luxas work from #78173 , please let me know if this is missing some functionality / tests !

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/assign @stealthybox 
/assign @sttts
/assign @lavalamp

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// encoding of an object, or an error.
+func (SimpleMetaFactory) Interpret(data []byte) (*schema.GroupVersionKind, error) {
+	findKind := struct {
+		// +optional
+		APIVersion string `json:"apiVersion,omitempty"`
+		// +optional
+		Kind string `json:"kind,omitempty"`
+	}{}
+	if err := yaml.Unmarshal(data, &findKind); err != nil {
+		return nil, fmt.Errorf("couldn't get version/kind; json parse error: %v", err)
```
Comment (neolit123) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go]: `return nil, fmt.Errorf("could not interpret GroupVersionKind; unmarshal error: %v", err)`

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+// SimpleMetaFactory provides default methods for retrieving the type and version of objects
+// that are identified with an "apiVersion" and "kind" fields in their JSON
+// serialization. It may be parameterized with the names of the fields in memory, or an
+// optional list of base structs to search for those fields in memory.
+type SimpleMetaFactory struct{}
+
+// Interpret will return the APIVersion and Kind of the JSON wire-format
+// encoding of an object, or an error.
+func (SimpleMetaFactory) Interpret(data []byte) (*schema.GroupVersionKind, error) {
+	findKind := struct {
```
Comment (neolit123) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go]: this should be a TypeMeta instance instead:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/types.go#L36-L41

`typeMeta := runtime.TypeMeta{}`

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	if gvk.Kind != expected.kind {
+		helperFailMsg(t, fmt.Sprintf("GVK mismatch of '%#v'", gvk), "", gvk.Kind)
+	}
+	if gvk.Version != expected.version {
+		helperFailMsg(t, fmt.Sprintf("GVK mismatch of '%#v'", gvk), "", gvk.Version)
+	}
+}
+
+func TestInterpret(t *testing.T) {
```
Comment (neolit123) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go]: the existing pattern for unit tests is slightly different:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version_test.go#L138-L161

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/scheme_test.go#L613

- define the test cases as a slice of structs as a inline element of the Test* function.
- all test cases should have names/descriptions - e.g. `name: successfully interpret GVK`.
- iterate over all test cases and call Interpret() tests on them.

**Review by neolit123 — COMMENTED** [review]
thanks for the PR @obitech 
i've added some comments.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/yaml"
+)
+
+// DefaultMetaFactory is a default factory for versioning objects in JSON. The object
+// in memory and in the default JSON serialization will use the "kind" and "apiVersion"
+// fields.
+var DefaultMetaFactory = SimpleMetaFactory{}
```
[the remark this answers is not in the document]
Comment (obitech) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go]: I adjusted the comment.

File: cmd/kubeadm/app/util/marshal.go [hunk 5]
```diff
@@ -92,31 +91,19 @@ func SplitYAMLDocuments(yamlBytes []byte) (map[schema.GroupVersionKind][]byte, e
...
-		if err := yaml.Unmarshal(b, &typeMetaInfo); err != nil {
-			return nil, err
-		}
-		// Require TypeMeta information to be present
-		if len(typeMetaInfo.APIVersion) == 0 || len(typeMetaInfo.Kind) == 0 {
-			errs = append(errs, errors.New("invalid configuration: kind and apiVersion is mandatory information that needs to be specified in all YAML documents"))
-			continue
+		gvk, err := yamlserializer.DefaultMetaFactory.Interpret(b)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid configuration: kind and apiVersion is mandatory information that needs to be specified in all YAML documents")
```
[the remark this answers is not in the document]
Comment (obitech) [reply cmd/kubeadm/app/util/marshal.go]: So you meant to leave the `schema.ParseGroupVersion` in `Interpret` but add an additional check in `SplitYAMLDocuments` ? I've done it like that now.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go (hunk 1 above)
Comment (obitech) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go]: Fixed.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go (hunk 2 above)
Comment (obitech) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go]: Fixed.

File: cmd/kubeadm/app/util/marshal.go [hunk 6]
```diff
@@ -80,7 +80,6 @@ func SplitYAMLDocuments(yamlBytes []byte) (map[schema.GroupVersionKind][]byte, e
 	buf := bytes.NewBuffer(yamlBytes)
 	reader := utilyaml.NewYAMLReader(bufio.NewReader(buf))
```
[the remark this answers is not in the document]
Comment (obitech) [reply cmd/kubeadm/app/util/marshal.go]: Ok maybe I can do this in a follow-up PR?

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go (hunk 3 above)
Comment (obitech) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go]: Ah thanks for pointing me towards that! I adjusted the test cases and I like them much better now.

File: cmd/kubeadm/app/util/marshal.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/kubeadm/app/util/marshal.go]: you can bring this in a WG component-standard meeting for discussion if you like.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+			input: `foo`,
+			errFn: func(err error) bool {
+				return err != nil
+			},
+		},
+	// 	TODO: Add JSON tests
+	}
+
+	for _, test := range testCases {
+		actual, err := DefaultMetaFactory.Interpret([]byte(test.input))
```
Comment (neolit123) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go]: please run these as sub-tests.
t.Run(...) as the examples i gave before.

File: cmd/kubeadm/app/util/marshal.go [hunk 8]
```diff
@@ -92,31 +91,23 @@ func SplitYAMLDocuments(yamlBytes []byte) (map[schema.GroupVersionKind][]byte, e
...
-		if err := yaml.Unmarshal(b, &typeMetaInfo); err != nil {
+		gvk, err := yamlserializer.DefaultMetaFactory.Interpret(b)
+		if err != nil {
 			return nil, err
 		}
-		// Require TypeMeta information to be present
-		if len(typeMetaInfo.APIVersion) == 0 || len(typeMetaInfo.Kind) == 0 {
-			errs = append(errs, errors.New("invalid configuration: kind and apiVersion is mandatory information that needs to be specified in all YAML documents"))
-			continue
+		if len(gvk.Version) == 0 || len(gvk.Kind) == 0 {
```
Comment (neolit123) [inline cmd/kubeadm/app/util/marshal.go]: please check gvk.Group too.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+		switch {
+		case test.errFn != nil:
+			if !test.errFn(err) {
+				t.Errorf("unexpected error returned with test '%s', got: %v", test.name, err)
```
Comment (neolit123) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go]: '%s' -> %q

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+			}
+		case err != nil:
+			t.Errorf("unexpected error with test '%s': %v", test.name, err)
```
Comment (neolit123) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go]: [1]

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+		case !reflect.DeepEqual(test.expected, actual):
+			t.Errorf("outcome mismatch with test '%s' -- expected %v, actual: %v",
```
Comment (neolit123) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go]: [1]

**Comment by neolit123** [comment]
thanks for the update, added some more comments.

**Review by rosti — COMMENTED** [review]
Thanks @obitech for picking this up!
Is there a doc or an issue, that explains in greater detail what we are trying to achieve here? This looks like the first step of a bigger effort and I wonder how this is supposed to continue.
Perhaps @luxas or @stealthybox have more info here?

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go (hunk 7 above)
Comment (obitech) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go]: Ah I missed that. Fixed now.

File: cmd/kubeadm/app/util/marshal.go (hunk 8 above)
Comment (obitech) [reply cmd/kubeadm/app/util/marshal.go]: Added.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go (hunk 9 above)
Comment (obitech) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go]: I included the name of the test as the actual test name in `t.Run()`, so I removed the print of the name here.

**Comment by neolit123** [comment]
@obitech 
forgot to mention, please squash the commits into one and force-push.

**Comment by obitech** [comment]
Squashed!

/test pull-kubernetes-kubemark-e2e-gce-big

**Comment by rosti** [comment]
@obitech , @stealthybox what's the plan for continuing this work in the future? Are we going to extract `SplitYAMLDocuments`? What's the end result, that we try to achieve here?

**Comment by obitech** [comment]
@rosti I've put this up as an agenda item for tomorrow's component-standard meeting where I would like to discuss the road map and get more information about the bigger picture. I'll update here afterwards.

**Comment by stealthybox** [comment]
[Tues Aug13 WG meeting notes](https://docs.google.com/document/d/18TsodX0fqQgViQ7HHUTAhiAwkf6bNhPXH4vNVTI7GwI/edit#heading=h.n3iz1rg9hvxt)

Would like to review other multi-doc support in existing code other than kubeadm
- Kubectl multi-doc
- Kustomize stuff
- Anything using splitYamlDocuments() in yaml decoder by proxy of an Exported function?
https://github.com/kubernetes/apimachinery/blame/master/pkg/util/yaml/decoder.go#L142

**Comment by luxas** [comment]
ping @sttts @liggitt @lavalamp @mtaufen  for review and approval.
@obitech needs a small autogenerated BUILD file rebase.
Otherwise this is ready to merge

**Comment by liggitt** [comment]
/approve

serializer/yaml change lgtm

**Comment by obitech** [comment]
Updated, rebased and squashed.

@luxas thanks for your comment on this! I believe there were still some questions in the discussion by @rosti what the future plans are for this effort. Since this started some time ago and I just carried it over, were there any ideas on how to build upon this patch to serialize multiple YAML files? Can we use this to be used inside a `CodecFactory`, for example?

**Comment by luxas** [comment]
> @rosti what the future plans are for this effort. Since this started some time ago and I just carried it over, were there any ideas on how to build upon this patch to serialize multiple YAML files? Can we use this to be used inside a CodecFactory, for example?

We still need it, but it's nothing that will be used by an "end-user" of API machinery. Instead, it'll be used internally for higher-level Serializer/CodecFactory-style abstractions that are able to e.g. read a YAML file and directly assess what type it is, **without** marshalling the full type into memory. You can take a look at my ConfigSerializer prototypes in earlier PRs showcasing this: #72883

**Comment by stealthybox** [comment]
Just talked about this in kubeadm triage.
This patch is looking good.

We recall there was some discussion of how generally useful this abstraction is.

Alex did some work to identify how multi-doc is handled in other places such as kubectl, and we found other points of interest in the multiple YAML Unmarshal code-paths that exist across the project.

@obitech, can we dig up that work before we decide to merge this?
I personally think this refactor is still a useful step toward working with multi-doc and ComponentConfig.

I'll schedule this for discussion on the next wg-component-standard call.

**Comment by obitech** [comment]
> @obitech, can we dig up that work before we decide to merge this?

I've documented some steps here: https://github.com/obitech/k8s-yaml-test

**Pull request merged**